New Transaction Signing and Fee Management

Gauthamastro opened the pull request:
Closes https://github.com/Polkadex-Substrate/thea/issues/3


NOTE: It will break the use of PolkadotJS explorer.

Gauthamastro committed:
Initial Commit

Gauthamastro committed:
Basic transaction payment setup

Gauthamastro committed:
Time to test

Gauthamastro committed:
its compiling

Gauthamastro committed:
Signed payload new extrinsic structure

Gauthamastro committed:
Remove lock

Gauthamastro committed:
Add Eth signing logic

Gauthamastro committed:
MM signing logic works now

Gauthamastro committed:
Modified Signature verification

Gauthamastro committed:
Tests are passing for runtime

Gauthamastro committed:
Extrinsics to add and remove tokens for fees

Gauthamastro committed:
Clippy and FMT

Gauthamastro committed:
Fix test

Gauthamastro committed:
Merge branch 'Develop' into gj-signed-payload

Gauthamastro committed:
Merge conflicts

Gauthamastro commented:
Tarpaulin is failing because of an older version of ethers library, it can only be upgraded by upgrading substrate version

Gauthamastro committed:
temp fix

nuel77 committed:
chore: revert back struct for signed payload and trip last byte for all transactions before signature validation

Gauthamastro committed:
revert signing Payload struct (#619)

Gauthamastro committed:
cargo fmt

nuel77 committed:
fix: removed trimming of signature on node side

serhii-temchenko committed:
Added tarpaulin ci flag to fix legacy error

serhii-temchenko commented:
> Tarpaulin is failing because of an older version of ethers library, it can only be upgraded by upgrading substrate version

@Gauthamastro please take a look on PR #621, which targeting this branch. It Includes fix for the tarpaulin issue.

Gauthamastro committed:
Fix of tarpaulin in the CI (#621)

Gauthamastro commented:
> > Tarpaulin is failing because of an older version of ethers library, it can only be upgraded by upgrading substrate version
> 
> 
> 
> @Gauthamastro please take a look on PR #621, which targeting this branch. It Includes fix for the tarpaulin issue.

Thanks for this!

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	/// Configure the pallet by specifying the parameters and types on which it depends.
```
Please remove one comment

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// A way to add new tokens for payment for fees
+		#[pallet::weight(10000)]
```
Any benchmarks should be added? Please add todo with reference to that issue to keep track of it.

35359595 replied:
this pallet is root only - some generic big enough value should suffice. Like 100_000_000_000

Gauthamastro replied:
Generic big values will make it too expensive for root. Proper benchmark is required.

nakul1010 replied:
Created [issue 666](https://github.com/Polkadex-Substrate/Polkadex/issues/666) to fix it.

serhii-temchenko replied:
Resolving this since it will be implemented in the scope of separate task.

serhii-temchenko replied:
For the info - benchmarks were implemented and merged in this branch from PR #671.

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+	impl<T: Config> Pallet<T> {
+		/// A way to add new tokens for payment for fees
+		#[pallet::weight(10000)]
+		pub fn allow_list_token_for_fees(
+			origin: OriginFor<T>,
+			asset: AssetIdOf<T>,
```
Please add params description into the dockblock

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+			origin: OriginFor<T>,
+			asset: AssetIdOf<T>,
+		) -> DispatchResult {
+			// TODO: Add a governance origin instead of root
```
Is there any task for that already created?

nakul1010 replied:
Fixed

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+
+		/// A way to remove tokens for payment for fees
+		#[pallet::weight(10000)]
+		pub fn block_token_for_fees(origin: OriginFor<T>, asset: AssetIdOf<T>) -> DispatchResult {
```
Please add params description into the dockblock

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+		len: usize,
+	) -> Result<(BalanceOf<T>, InitialPayment<T>), TransactionValidityError> {
+		let fee = pallet_transaction_payment::Pallet::<T>::compute_fee(len as u32, info, self.tip);
+		debug_assert!(self.tip <= fee, "tip should be included in the computed fee");
```
From the documentation:
```rust
/// Unlike [`assert!`], `debug_assert!` statements are only enabled in non
/// optimized builds by default. An optimized build will not execute
/// `debug_assert!` statements unless `-C debug-assertions` is passed to the
/// compiler. This makes `debug_assert!` useful for checks that are too
/// expensive to be present in a release build but may be helpful during
/// development.
```
It will not be included in release build. Is it intentional? I assume there should be proper assertion with handling error without panicking?

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+		if fee.is_zero() {
+			Ok((fee, InitialPayment::Nothing))
+		} else if self.asset_id != Zero::zero() {
+			// If the asset id is zero, then we treat that case as payment in PDEX,
```
Is it missed part in comment? because in `else if` block checked that asset id is not zero, but comment says `If the asset id is zero`

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+		len: usize,
+	) -> TransactionValidity {
+		use pallet_transaction_payment::ChargeTransactionPayment;
+		let (fee, inital_payment) = self.withdraw_fee(who, call, info, len)?;
```
typo in `inital_payment`

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+// Asset id type alias.
+pub(crate) type ChargeAssetIdOf<T> =
+	<<T as Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::AssetId;
+// Liquity info type alias.
```
typo in `Liquity`

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+	<T as pallet_transaction_payment::Config>::OnChargeTransaction;
+// Balance type alias.
+pub(crate) type BalanceOf<T> = <OnChargeTransactionOf<T> as OnChargeTransaction<T>>::Balance;
+// Liquity info type alias.
```
typo in `Liquity`

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let (_fee, initial_payment) = self.withdraw_fee(who, call, info, len)?;
+		// Check if the given asset is valid
+		if let InitialPayment::Asset(asset) = &initial_payment {
```
Shouldn't be returner error in case of `else`?

35359595 replied:
Should Emmitt error event

35359595 replied:
Or at least log

nakul1010 replied:
Passed an error back to caller fn

serhii-temchenko replied:
Which caller function? This function should be safe to call without outer scope. Assertion/error should be returned if its not valid since its even not validates asset because of this if...

serhii-temchenko commented on pallets/assets-transaction-payment/src/lib.rs:
```diff
@@ -0,0 +1,335 @@
+					// non-zero here in case of unsigned extrinsics as they don't pay fees but
+					// `compute_actual_fee` is not aware of them. In both cases it's fine to just
+					// move ahead without adjusting the fee, though, so we do nothing.
+					debug_assert!(tip.is_zero(), "tip should be zero if initial fee was zero.");
```
Same as above

35359595 replied:
Emmitt event or log

nakul1010 replied:
Passed an error back to caller fn

serhii-temchenko replied:
Logic here changed and assertion removed at all. Here is a new implementation by @nakul1010:

```rust
InitialPayment::Nothing => {
	// `actual_fee` should be zero here for any signed extrinsic. It would be
	// non-zero here in case of unsigned extrinsics as they don't pay fees but
	// `compute_actual_fee` is not aware of them. In both cases it's fine to just
	// move ahead without adjusting the fee, though, so we do nothing.
	return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
},

```

Which is really different from the initial implementation from @Gauthamastro with assertion and is not match comment description. Is it was discussed and decided to be changed to return error here? If so - please update the comment. If not - add assertion that `tip.is_zero()` and if not - only then return error.

Gauthamastro committed:
fix: removed trimming of signature on node side (#620)

ix: removed trimming of payload on node side for signature verification

Gauthamastro committed:
cargo fmt

serhii-temchenko reviewed (changes requested):
Are there any plans to extend tests for this? If so do we already have an issue for this?

Gauthamastro committed:
Merge branch 'Develop' into gj-signed-payload

serhii-temchenko commented on .github/workflows/ci.yml:
```diff
@@ -87,7 +87,8 @@ jobs:
         with:
-          args: '--avoid-cfg-tarpaulin --exclude polkadex-node node-polkadex-runtime --workspace --timeout 180'
+          version: '0.22.0'
+          args: '--exclude polkadex-node node-polkadex-runtime --workspace --timeout 180'
```
I assume conflict was resolved without leaving `--avoid-cfg-tarpaulin` arg. Will raise PR to dev branch with that fix.

Gauthamastro replied:
Okay thanks

serhii-temchenko replied:
[Here](https://github.com/Polkadex-Substrate/Polkadex/pull/630) is a PR.

nakul1010 committed:
mock environment

nakul1010 committed:
added trait implementation in mock

nakul1010 committed:
added crate for transactions

nakul1010 committed:
added in toml file

nakul1010 committed:
updated client

nakul1010 committed:
feature gated

nakul1010 committed:
added Unchecked Extrinsic type

nakul1010 committed:
Mock block and unchecked extrinsic

nakul1010 committed:
added treasury pallet

nakul1010 committed:
removed unused types

35359595 commented on node/Cargo.toml:
```diff
@@ -99,6 +99,7 @@ sp-trie = { git = "https://github.com/paritytech/substrate", branch = "polkadot-
 
 frame-try-runtime = { git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.28", optional = true }
 try-runtime-cli = { git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.28", optional = true }
+assets-transaction-payment = {path = "../pallets/assets-transaction-payment"}
```
preserve formatting with white spaces

35359595 commented on pallets/assets-transaction-payment/Cargo.toml:
```diff
@@ -0,0 +1,49 @@
+description = "FRAME pallet for migrating ERC20 PDEX to Native PDEX"
+edition = "2021"
+homepage = "https://polkadex.trade"
+name = "assets-transaction-payment"
```
all pallets should prefix with 'pallet-' for better readability

35359595 commented on pallets/assets-transaction-payment/Cargo.toml:
```diff
@@ -0,0 +1,49 @@
+scale-info = { workspace = true, default-features = false, features = ["derive"] }
+
+frame-support = { workspace = true, default-features = false }
+frame-benchmarking = { workspace = true, default-features = false,  optional = true }
```
benchmarking is not implemented, but imported

35359595 commented on pallets/assets-transaction-payment/src/payment.rs:
```diff
@@ -0,0 +1,177 @@
+		let can_withdraw =
+			<T::Fungibles as Inspect<T::AccountId>>::can_withdraw(asset_id, who, converted_fee);
+		if !matches!(can_withdraw, WithdrawConsequence::Success) {
+			return Err(InvalidTransaction::Payment.into())
```
This same error is used in three different cases with different actions, which might complicate troubleshooting. Should we dedicated errors instead?

nakul1010 replied:
The code is taken from substrate `transaction-pallet` so I don't think changing it's a good idea.

serhii-temchenko replied:
Agree with @35359595 here. Even if its taken from someone is not saying that its implemented correct by default... if we see some problem in logic - it should be improved...

35359595 commented on pallets/assets-transaction-payment/src/payment.rs:
```diff
@@ -0,0 +1,177 @@
+		_tip: Self::Balance,
+		paid: Self::LiquidityInfo,
+	) -> Result<(), TransactionValidityError> {
+		let min_converted_fee = if corrected_fee.is_zero() { Zero::zero() } else { One::one() };
```
This code is repeated in previous function. Please extract into separate function.

nakul1010 replied:
The code is taken from substrate transaction-pallet so I don't think changing it's a good idea.

35359595 commented on pallets/assets-transaction-payment/src/payment.rs:
```diff
@@ -0,0 +1,177 @@
+	) -> Result<(), TransactionValidityError> {
+		let min_converted_fee = if corrected_fee.is_zero() { Zero::zero() } else { One::one() };
+		// Convert the corrected fee into the asset used for payment
+		let converted_fee = CON::to_asset_balance(corrected_fee, paid.asset())
```
Above can include this bit of code as well

nakul1010 replied:
The code is taken from substrate transaction-pallet so I don't think changing it's a good idea.

35359595 commented on runtime/Cargo.toml:
```diff
@@ -100,14 +101,26 @@ polkadex-primitives = { workspace = true, default-features = false }
 thea = {path = "../pallets/thea", default-features = false }
 thea-primitives = { path = "../thea-primitives", default-features = false  }
 
+
```
Double empty line

35359595 commented on runtime/Cargo.toml:
```diff
@@ -186,10 +208,12 @@ std = [
     "pallet-nomination-pools/std",
     "pallet-asset-handler-runtime-api/std",
     "thea/std",
+
```
Remove this empty line in a list

35359595 commented on runtime/src/lib.rs:
```diff
@@ -92,6 +92,10 @@ pub mod impls;
 pub mod constants;
+pub mod ethereum;
+mod signature;
+pub mod signedpayload;
```
Can this name use underscore as word separator?

nakul1010 replied:
used `signed_payload`

35359595 commented on runtime/src/lib.rs:
```diff
@@ -1295,6 +1305,25 @@ impl asset_handler::pallet::Config for Runtime {
+pub struct AlternateTokenSwapper;
+impl HandleSwap<Runtime> for AlternateTokenSwapper {
+	fn swap(credit: CreditOf<AccountId, Assets>) -> NegativeImbalanceOf<Runtime> {
+		// TODO: @Emmanuel integrate Polkapool here
```
Is this been planned in some issue or worked on? Should it be implemented before merge?

nakul1010 replied:
@nuel77 can u comment on this

35359595 commented on runtime/src/signature.rs:
```diff
@@ -0,0 +1,42 @@
+		match (&self.signature, signer) {
+			(Signature::Ed25519(ref sig), who) => match ed25519::Public::from_slice(who.as_ref()) {
+				Ok(signer) => sig.verify(msg, &signer),
+				Err(()) => false,
```
Should false cases log to simplify debugging?

nakul1010 replied:
logged it

35359595 commented on runtime/src/signedpayload.rs:
```diff
@@ -0,0 +1,84 @@
+
+pub fn substrate_signing(payload: &[u8]) -> Vec<u8> {
+	if payload.len() > 256 {
+		blake2_256(payload).to_vec()
```
Can we take payload as impl AsRef<[u8]> and return slice of it to avoid allocation?

Gauthamastro replied:
That's a good idea.

35359595 commented on runtime/src/unchecked_extrinsic.rs:
```diff
@@ -0,0 +1,437 @@
+
+	type SignaturePayload = (Address, Signature, Extra);
+
+	fn is_signed(&self) -> Option<bool> {
```
This is antipattern. Can we use enum for return type here?

Gauthamastro replied:
That trait is controlled by substrate.

35359595 commented on runtime/src/unchecked_extrinsic.rs:
```diff
@@ -0,0 +1,437 @@
+		match self.signature.as_ref() {
+			Some(s) => {
+				tmp.push(EXTRINSIC_FORMAT_VERSION | 0b1000_0000);
+				s.encode_to(&mut tmp);
```
Are we ok that allocated tmp will be actually smaller if 's' in None?

Gauthamastro replied:
I think it will fine.

35359595 commented on runtime/src/unchecked_extrinsic.rs:
```diff
@@ -0,0 +1,437 @@
+	{
+		let r = sp_core::bytes::deserialize(de)?;
+		Decode::decode(&mut &r[..])
+			.map_err(|e| serde::de::Error::custom(format!("Decode error: {e}")))
```
If we use 'custom(e.to_string())' we can drop extra allocation from 'format!'

Gauthamastro replied:
yup, @nakul1010

nakul1010 committed:
test cases for extrinsic added

nakul1010 committed:
added more tc

nakul1010 committed:
transaction_payment_in_native_possible TC added

nakul1010 committed:
test case added without asserting balance

nakul1010 committed:
added assertions

nakul1010 committed:
fixed fmt

nakul1010 committed:
fixed clippy

nakul1010 committed:
added fmt

nakul1010 committed:
fixed comment

nakul1010 committed:
fixed clippy

nakul1010 committed:
allow clippy

nakul1010 committed:
added fmt

nakul1010 committed:
allow clippy warning for enum

nakul1010 committed:
allow clippy for entire file

nakul1010 committed:
fixed ci

nakul1010 committed:
pinn rust version

serhii-temchenko committed:
Merge branch 'Develop' into gj-signed-payload

# Conflicts:
#	Cargo.lock
#	runtime/Cargo.toml
#	runtime/src/lib.rs

serhii-temchenko committed:
Revert "pinn rust version"

This reverts commit d059201cc849008bd5124e6ca95f5b61c431f9f3.

serhii-temchenko committed:
Revert "fixed ci"

This reverts commit da67f8bee9b18ad9c8e60819da950e58dcc50e1e.

serhii-temchenko committed:
Revert "allow clippy for entire file"

This reverts commit 54d411c19ad3ae88b9828a2fc4f4a6a192dc2a9f.

serhii-temchenko committed:
Revert "allow clippy warning for enum"

This reverts commit e5607c97efbcec016fe7fe471d3e4610feb8d07a.

serhii-temchenko committed:
Revert "added fmt"

This reverts commit 7c8689ed7473897451e2240528c58927ad54def8.

serhii-temchenko committed:
Revert "allow clippy"

This reverts commit 3d9fe252001c55cb096baded7ef7664e47dac72f.

serhii-temchenko committed:
Revert "fixed clippy"

This reverts commit 5453a81183364eb704268163d15b5e00cf214c36.

serhii-temchenko committed:
Merge branch 'gj-signed-payload' into nap-testcase-signed-payload

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	runtime/Cargo.toml

nakul1010 committed:
fixed reviews

nakul1010 committed:
Merge remote-tracking branch 'origin/nap-testcase-signed-payload' into nap-testcase-signed-payload

nakul1010 committed:
changed crate name

nakul1010 committed:
fmt added

nakul1010 committed:
Additions for 616 (#638)

closes #652.

nakul1010 committed:
fix reviews

nakul1010 committed:
added pallet prefix for better readability

nakul1010 committed:
removed frame-benchmarking dependency.

nakul1010 committed:
Merge branch 'Develop' of https://github.com/Polkadex-Substrate/Polkadex into gj-signed-payload

# Conflicts:
#	Cargo.lock
#	runtime/Cargo.toml
#	runtime/src/lib.rs

nakul1010 committed:
handled conflicts

nakul1010 committed:
removed unused dependency

nakul1010 committed:
fmt added

nakul1010 committed:
fix TC

nakul1010 committed:
fix TC

serhii-temchenko reviewed (changes requested):
@nakul1010 there is a list of unresolved conversations and some new comments on resolved comments by you. Please take a look on them and return to re-review after the all will be resolved.

35359595 committed:
assets-transaction-payment benchmarks

serhii-temchenko committed:
assets-transaction-payment benchmarks (#671)

## Describe your changes
benchmarks and weights generated on dedicated HW machine for
pallet-assets-transaction-payment

## Issue ticket number and link
#666 

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [x] I removed all Clippy and Formatting Warnings. 
- [x] I added required Copyrights.

Gauthamastro committed:
Merge branch 'Develop' into gj-signed-payload

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	client/Cargo.toml
#	client/src/lib.rs
#	node/Cargo.toml
#	runtime/Cargo.toml
#	runtime/src/lib.rs

Gauthamastro committed:
somewhere

Gauthamastro committed:
Merge branch 'Develop' into gj-signed-payload

# Conflicts:
#	Cargo.lock

Gauthamastro committed:
upgrades stuff to 37

Gauthamastro committed:
Merge branch 'Develop' into gj-signed-payload

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	check-all-ci-tests.sh
#	client/Cargo.toml
#	client/src/lib.rs
#	node/Cargo.toml
#	runtime/Cargo.toml
#	runtime/src/lib.rs

Gauthamastro committed:
wallet connector is compiling
